torch_xla/csrc/tensor_methods.cpp

```diff
@@ -1537,11 +1537,19 @@ XLATensorPtr lt(const XLATensorPtr& input, const XLATensorPtr& other) {
   return DispatchComparisonOp(at::aten::lt, input, other);
 }
 
-void masked_fill_(XLATensorPtr& input, const XLATensorPtr& mask,
-                  const at::Scalar& value) {
+XLATensorPtr masked_fill(XLATensorPtr& input, const XLATensorPtr& mask,
+                         const at::Scalar& value) {
   torch::lazy::ScopePusher ir_scope(at::aten::masked_fill.toQualString());
-  input->SetIrValue(torch::lazy::MakeNode<MaskedFill>(
-      input->GetIrValue(), MaybeExpand(mask->GetIrValue(), input->shape()),
+  auto input_value = input->GetIrValue();
+  // Expand input tensor to mask if needed (same as masked_scatter below).
+  // An additional check makes sure to only expand if the rank of input tensor
+  // is less than that of the mask tensor.
+  if (input->shape().get().rank() < mask->shape().get().rank() &&
+      input->shape().get().dimensions() < mask->shape().get().dimensions()) {
```

Comment:
This part may require a little bit of additional context. The 2nd part of the if statement `input->shape().get().dimensions() < mask->shape().get().dimensions()` is the same as https://github.com/pytorch/xla/pull/4438. 

The first part of the if statement `input->shape().get().rank() < mask->shape().get().rank()` only applies to the `masked_fill` ops.  This is to handle cases that is covered by one of our existing cpp tests `MaskedFillBroadcast`, with shapes like:
```
self = {2, 5, 4, 3}
mask {4, 1}
```
In this specific set of input shapes, the 2nd if statement yields to true because mask[0] is greater than self[0]. To prevent this, we are add the 1st part of the if statement to make sure to do the comparison for expanding only if the rank of input tensor is less than that of the mask tensor.

In the previous PR https://github.com/pytorch/xla/pull/4438, this wasn't needed because the `masked_scatter` op doesn't allow such expansion. Ex:
```
>>> self = torch.tensor([3, -5, 3, -2, 5])
>>> mask = torch.tensor([[True, False, True, False, False],[False, False, True, False, False],[True, False, False, False, True],[True, True, False, True, False],[False, True, False, True, True]])
>>> value = torch.tensor(-3)
>>> 
>>> torch.masked_scatter(self, mask, value)
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
RuntimeError: Number of elements of source < number of ones in mask
```
However, the `masked_fill` allows such expansion hence these checks are required:
```
>>> self = torch.tensor([3, -5, 3, -2, 5])
>>> mask = torch.tensor([[True, False, True, False, False],[False, False, True, False, False],[True, False, False, False, True],[True, True, False, True, False],[False, True, False, True, True]])
>>> value = torch.tensor(-3)
>>> self.masked_fill(mask, value)
tensor([[-3, -5, -3, -2,  5],
        [ 3, -5, -3, -2,  5],
        [-3, -5,  3, -2, -3],
        [-3, -3,  3, -3,  5],
        [ 3, -3,  3, -3, -3]])
```

Reply:
One question: is the second condition still needed in this case?
```
input->shape().get().dimensions() < mask->shape().get().dimensions()
```

Reply:
Yep, to cover cases like this:
```
>>> self = torch.tensor([[2], [-3]])
>>> mask = torch.tensor([[True, False, True],[False, False, True]])
>>> value = torch.tensor(-3)
>>> self.shape
torch.Size([2, 1])
>>> mask.shape
torch.Size([2, 3])
>>> self.masked_fill(mask, value)
tensor([[-3,  2, -3],
        [-3, -3, -3]])
```
The rank of two tensors are equal, but we still need to expand the self tensor.

Reply:
Not sure the case you just mentioned is included in the current code.

Reply:
You're absolutely right, thanks for the catch. What I needed to do was: `input->shape().get().rank() <= mask->shape().get().rank()`. Updated. Also added another broadcast cpp test to cover this scenario.